backend/backendutil/bodystructure.go

```diff
@@ -48,14 +73,41 @@ func FetchBodyStructure(header textproto.Header, body io.Reader, extended bool)
 			parts = append(parts, pbs)
 		}
 		bs.Parts = parts
+	} else {
+		countedBody := countReader{r: body}
+		needLines := false
+		if bs.MIMEType == "message" && bs.MIMESubType == "rfc822" {
+			// This will result in double-buffering if body is already a
+			// bufio.Reader (most likely it is). :\
+			bufBody := bufio.NewReader(&countedBody)
+			subMsgHdr, err := textproto.ReadHeader(bufBody)
+			if err != nil {
+				return nil, err
+			}
+			bs.Envelope, err = FetchEnvelope(subMsgHdr)
+			if err != nil {
+				return nil, err
+			}
+			// TODO: Limit recursion.
```

Comment:
I wonder if we should do this in go-message directly. Would probably be a good idea.

Reply:
I will drop that TODO, I think we have an issue open in go-message already.